ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,168 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module XenAPI = Client.Client
+module Date = Xapi_stdext_date.Date
+
+type message_name_t = string
+
+type message_priority_t = int64
+
+type message_id_t = message_name_t * message_priority_t
+
+type remaining_days_t = int
+
+type expiry_messaging_info_t = {
+    message_cls:
+      [ `Certificate
+      | `Host
+      | `PVS_proxy
+      | `Pool
+      | `SR
+      | `VDI
+      | `VM
+      | `VMPP
+      | `VMSS ]
+  ; message_obj_uuid: string
+  ; obj_description: string
+  ; message_sent_on_remaining_days_list: (remaining_days_t * message_id_t) list
+  ; expiry: Xapi_stdext_date.Date.t (* when the obj will expire *)
+}
+
+let seconds_per_day = 3600. *. 24.
+
+let days_until_expiry epoch expiry =
+  (expiry /. seconds_per_day) -. (epoch /. seconds_per_day)
+
+let all_messages rpc session_id =
+  XenAPI.Message.get_all_records ~rpc ~session_id
+
+let related_messages msg_obj_uuid related_message_name_list all_msgs =
+  List.filter
+    (fun (_ref, record) ->
+      let related name = List.mem name related_message_name_list in
+      record.API.message_obj_uuid = msg_obj_uuid
+      && related record.API.message_name
+    )
+    all_msgs
+
+let message_body msg expiry =
+  let valid_till = Date.of_float (Date.to_float expiry -. 1.) in
+  Printf.sprintf "<body><message>%s</message><date>%s</date></body>" msg
+    (Date.to_string valid_till)
+
+let expired_message obj = Printf.sprintf "The %s has expired." obj
+
+let expiring_message obj = Printf.sprintf "The %s is expiring soon." obj
+
+let generate_alert now obj_description message_sent_on_remaining_days_list
+    expiry =
+  let remaining_days =
+    days_until_expiry (Date.to_float now) (Date.to_float expiry)
+  in
+  let rec get_most_critical = function
```

Comment:
> If you assume that the list is already sorted, which the certificate one is, and which you can mandate in the interface
Yes, if the list is already sorted, it is easier here. But then I think we will need to add some code in the interface to verify if it is sorted?

Reply:
It's an internal library used only by our own code, and doesn't take user input, so I think it's fine. Otherwise, just sort the list with e.g. `List.sort (fun (a, _) (b, _) -> compare a b)`.

Reply:
I think sorting the list is not necessary. What we want is something like finding the maximum value.

Reply:
I guess what @robhoes cares most is correctness, performance is the 2nd one.

Reply:
Changed back to use `List.sort` and `List.find_opt` instead of self-written func: `get_most_critical`.